File: ios/YeterNative.m
//
//  YeterNative.m
//  yeterApp
//
//  Created by Erdem Özkök on 9.02.2024.
//

#import <Foundation/Foundation.h>
#import "React/RCTBridgeModule.h"
#import <React/RCTEventEmitter.h>
#import <React/RCTViewManager.h>

@interface RCT_EXTERN_MODULE(YeterNative, RCTEventEmitter)

RCT_EXTERN_METHOD(addWord:(NSString *)word
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getWords:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(removeWord:(NSString *)word
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

@end

File: ios/YeterNative.swift
import Foundation
import React

@objc(YeterNative)
class YeterNative: RCTEventEmitter {
  
    public static var shared:YeterNative?
    private let userDefaults = UserDefaults(suiteName: "group.com.ozkokdev.yeter.appgroup")!

    override init() {
        super.init()
        YeterNative.shared = self
    }

    @objc(addWord:withResolver:withRejecter:)
    func addWord(word: NSString, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        var wordsList = userDefaults.array(forKey: "words") as? [String] ?? []
        wordsList.append(word as String)
        userDefaults.set(wordsList, forKey: "words")
        resolve(wordsList)
    }

    @objc(getWords:withRejecter:)
    func getWords(resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        let wordsList = userDefaults.array(forKey: "words") as? [String] ?? []
        resolve(wordsList)
    }

    @objc(removeWord:withResolver:withRejecter:)
    func removeWord(word: NSString, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        var wordsList = userDefaults.array(forKey: "words") as? [String] ?? []
        wordsList.removeAll { $0 == word as String }
        userDefaults.set(wordsList, forKey: "words")
        resolve(wordsList)
    }
}
